### ISSUE TITLE
Spyder-Like variable explorer

### ISSUE THREAD

**Issue by YashKarthik** [microsoft/vscode-jupyter#1148]
# Feature: Variable explorer 

The variable explorer in the notebooks is a important part of many data science IDEs, though vscode is not a full blown IDE I prefer it over others, its amazing. Vscode has a amazing features, just that the variable explorer doesn't feel/look so good, an explorer like in spyder or even Pycharm would be a nice feature and I'm sure many ppl would like that interface.
Thanks

**Comment by Ikhwansong** [comment]
I agree with them. So please improve the 'variable explorer'

**Comment by NikilXYZ** [comment]
I agree - this is especially useful for matrices and csv / table type data.
The one reason I use spyder (if I'm using Python) or use R studio (if I'm using R) instead of VS Code for scientific computing or image processing.

**Comment by johlo22** [comment]
I agree, I'd love to be able to use VS Code for scientific python since it's far superior to Spyder in other aspects but a working variable explorer and data viewer is an absolute requirement for a scientific IDE. The current VS Code one has too limited capabilities and also is far too slow to be useful.

**Comment by jbiz25** [comment]
@DonJayamanne @DavidKutu  Hi guys, just to expand upon the request.... It would be great if the focus wasn't only on "data viewing".  yes, that's important, but a more general, all-purpose variable viewer would be useful, where users could see all attributes and methods of each object alive in the kernel.  This is what Spyder offers.  It's not just a viewer for data objects.... it's a viewer for ALL variables, not matter their type.  I believe that Spyder pickles objects in order to view their attrs and methods in the Variable Explorer.  You can double click on any object and keep descending into the hierarchy of the objects inside of them.  And, regarding data viewing, currently complex numbers are not supported in the Data Viewer.  This is a requirement for my field of engineering.

**Comment by NikilXYZ** [comment]
Yep I agree with this. Also make it so the viewer refreshes the values on
any update to the variable. Currently in VS Code I have to keep opening a
new tab every time the values change

**Comment by jbiz25** [comment]
That's exactly it.  Will upvote.  Thanks @DonJayamanne

**Comment by mho21** [comment]
Thank you for picking up! as this drawback of vscode is the only reason I still use intellij right now, upvote both issue! #5532

**Comment by kickbone** [comment]
Could not agree more. Looking forward to an upgraded variable explorer in VS Code!

**Comment by IsaiahvH** [comment]
As far as I know, the current variable explorer also does not offer a way to search for a variable by name.
When there is many variables, scrolling in a small window becomes tedious and it would be great if this feature would also be considered.

**Comment by Y-H-Joe** [comment]
another vote here!

**Comment by Aizenz** [comment]
Vote for a better variable explorer too.
The color indicating the size of variable is really helpful!

**Comment by ldorigo** [comment]
+1. Also, I'm wondering if you couldn't "just" re-use the variable viewer that is used in the debugger? I find it 10x better in terms of interactivity than the current "JUPYTER: VARIABLES" pane.

**Comment by mho21** [comment]
> +1. Also, I'm wondering if you couldn't "just" **re-use the variable viewer that is used in the debugger**? I find it 10x better in terms of interactivity than the current "JUPYTER: VARIABLES" pane.

indeed, 90% I am starting the debugger only to view variables in better UI as a workaround - debuggers viewer is perfect!
less than 10% goes out to real debugging

**Comment by NikilXYZ** [comment]
But the variable debugger doesn't show arrays and matrices very cleanly.
Compare it to Spyder's viewer where the matrix actually shows up like a
grid of rectangles, and a 3D matrix has sort of a page drop down.

**Comment by cnw21** [comment]
I've tried to make the switch to VS Code like 20 times but every time I end up going back to Spyder. The variable explorer is amazing, and iPython being the console, being able to run any code for graphs just by highlighting a few lines of code and pressing f9, where it not only pops up in the console but stores it under plots. All great features its hard to get away from. If they could just make VS Code a little more convenient for data science in Python I'd make the jump.

**Comment by mack-raymond** [comment]
@cnw21 I've just switched to VScode from Spyder for most tasks. The interactive window has almost the same functionality as the ipython console in spyder. Highlight a code snippet in the editor and "shift+ enter" it into the interactive window. Plots can be rendered inline or in a separate window

**Comment by JOPEDSAN** [comment]
I agree with everything that is said in the issue. The reason why I don't definitely switch from Spyder to VS Code is the variable browser. I think it is essential when creating a structure to be able to visualize it and navigate with a double-click (example: dictionary of dictionaries, etc...). I hope that this improvement in the visualization of variables will be added, as soon as it will be solved it will be the definitive tool. 
Thanks !!!

**Comment by phgmacedo** [comment]
yes, spyder, pycharm, and matlab all have great variable and plot viewers

**Comment by MDK413** [comment]
I reached out to these comments through google from same but I can not get an answer.
Is there any update coming regarding that?

**Comment by phgmacedo** [comment]
> I reached out to these comments through google from same but I can not get an answer. Is there any update coming regarding that?

@MDK413 Like I said in #10998:

> With the upvotes from https://github.com/microsoft/vscode-jupyter/issues/1148 and https://github.com/microsoft/vscode-jupyter/issues/1138 it would be enough to make this issue the third most requested feature.

**Comment by KangOxford** [comment]
Please add a variable explorer!

**Comment by DonJayamanne** [comment]
More suggestions from here https://github.com/microsoft/vscode-jupyter/issues/11679
* showing a dataframe is fast
* it is opened in its own window that you can enlarge at will
* all columns are much wider and they can be even resized automatically

**Comment by Cglezf** [comment]
Please add a variable browser!!!

**Comment by mateuszrenno** [comment]
I agree! Spyder variable explorer is the best.

**Comment by jmew** [comment]
Just wanted to share this extension from Microsoft that integrates with Notebooks in VS Code, to allow you to open any Pandas DataFrame in an interactive data wrangler

https://marketplace.visualstudio.com/items?itemName=ms-toolsai.datawrangler

![image](https://github.com/microsoft/vscode-jupyter/assets/15910920/4729cbee-f63d-4982-b7fc-5b6d6d3be897)

![image](https://github.com/microsoft/vscode-jupyter/assets/15910920/805d865d-fd1d-4eb9-873e-452296e3a9eb)

**Comment by starball5** [comment]
What's the relationship between this issue ticket and the work being done with https://github.com/microsoft/vscode/issues/165445?

**Comment by BENSONCHIEN** [comment]
After almost 4 years, another vote here! 

Because vscode github copliot and Spyder variable explorer..., I keep switching between the two IDEs.

**Comment by wrgrant** [comment]
Nice @amunger !!
This is a great start and covers everything I need for script development.

**Comment by NikilXYZ** [comment]
Does this only work for notebook files?

*NIKIL RAGAV** | **CEO | **inventXYZ *

alum of Wharton Business | Penn Electrical Engineering
832.910.9645 | inventXYZ.com <http://www.inventxyz.com/>
Integrating hands-on real-world STEM & CS experiences into every core
middle & high school course

**Comment by WhiteRavenTechnology** [comment]
> you can try out the experimental built-in variable view with `"notebook.experimental.variablesView": true` - A "Notebook Variables" section should show up in the run/debug sidebar when a python kernel from the jupyter extension is running. It still needs some polish, but should be closer to what is asked for here.

Well I thought this was gonna be it. Can't view dataframes in any kind of dataviewer, and apparently when a variable becomes sufficiently complex (dictionary with nested dataframes), the "Notebook Variables" section just empties itself and stops working. Can't figure out what it is about my specific code that's causing it. The "Jupyter" variables tab is more useful than this since it at least doesn't empty itself.

Interestingly, when debugging, the normal variable viewer DOES work fine and allow me to dig into my dictionary objects and view dataframes in a data viewer from there (of course it says the dataviewer is going to be deprecated in favor of DataWrangler because why would we want something native that works instead of a bloated extension that doesn't).

I just want to be able to expand my dictionaries and view the dataframes contained on them. PyCharm does it. Spyder apparently does it. VS Code ALMOST does it perfectly when debugging. 

This is nuts.

**Comment by BrazilForever11** [comment]
How can I try `notebook.experimental.variablesView`? I pasted it into search bar, but nothing came back.

**Comment by BrazilForever11** [comment]
I managed to get it work `notebook.experimental.variablesView`  with Data Wrangler extension. I guess here is the bug tracker: https://github.com/microsoft/vscode-jupyter/pull/15141

This seems to work in notebook. Would it be possible to have it in flat python file?

**Comment by ejuliaschulz** [comment]
> How can I try `notebook.experimental.variablesView`? I pasted it into search bar, but nothing came back.

Click on File --> Preferences --> Settings (Or just Ctrl + ,), in the search bar of the window that opened up you can type `notebook.experimental.variablesView`, this will show the respective setting.

This is a great step forward, thank you @amunger ! 
Would be amazing to have it for flat python files as well.

**Comment by amunger** [comment]
The setting ID was updated to just `notebook.variablesView`, since at some point it won't be experimental, but the old setting key will still work.

The new variable view is available for plain python if you use the [native python REPL](https://code.visualstudio.com/docs/python/run#_native-repl) by setting `python.REPL.sendToNativeREPL` to true, or the [Interactive Window](https://code.visualstudio.com/docs/python/jupyter-support-py) from the jupyter extension if you want to work with jupyter kernels.
